scipy/optimize/_trustregion_exact.py

```diff
@@ -0,0 +1,425 @@
+
+    Returns
+    -------
+    delta : float
+        Amout that should be added to the element (k, k) of the
+        leading k by k submatrix of ``a`` to make it singular.
+    v : ndarray
+        A vector such that ``v.T B v = 0``. Where B is the matrix A after
+        ``delta`` is added to its element (k, k).
+    """
+
+    # Compute delta
+    delta = np.sum(U[:k-1, k-1]**2) - A[k-1, k-1]
+
+    # Get dimension
+    n = len(A)
+
+    # Inicialize v
+    v = np.zeros(n)
+    v[k-1] = 1
+
+    # Compute the remaining values of v by solving a triangular system
+    if k != 1:
+        v[:k-1] = solve_triangular(U[:k-1, :k-1], -U[:k-1, k-1])
+
+    return delta, v
+
+
+class IterativeSubproblem(BaseQuadraticSubproblem):
+    """Quadratic subproblem solved by nearly exact iterative method.
+
+    Notes
+    -----
+    This subproblem solver was based on _[1], _[2] and _[3],
+    which implement similar algorithms. The algorithm is basically
+    that of _[1] but ideas from _[2] and _[3] were also used.
+
+    References
+    ----------
+    .. [1] Conn, A. R., Gould, N. I., & Toint, P. L. (2000).
+           "Trust region methods". Siam. pp. 169-200.
+    .. [2] Nocedal, J., and  Wright, S. (2006).
+           "Numerical optimization". Springer Science &
+           Business Media. pp. 83-91.
+    .. [3] Moré, J. J., & Sorensen, D. C. (1983).
+           Computing a trust region step. SIAM Journal on Scientific
+           and Statistical Computing, 4(3), 553-572.
+    """
+
+    UPDATE_COEF = 0.01
+    EPS = np.finfo(float).eps
+
+    def __init__(self, x, fun, jac, hess, hessp=None,
+                 k_easy=0.1, k_hard=0.2):
+
+        # ``k_easy`` and ``k_hard`` are parameters used
+        # to detemine the stop criteria to the iterative
+        # subproblem solver. Take a look at pp. 194-197
+        # from reference _[1] for a more detailed description.
+
+        self.lb = None
+        self.previous_trust_radius = -1
+        self.niter = 0
+        self.k_easy = k_easy
+        self.k_hard = k_hard
+
+        # Superclass __init__
+        BaseQuadraticSubproblem.__init__(self, x, fun, jac,
+                                         hess)
+
+        # Get Lapack function for cholesky decomposition
+        self.cholesky, = get_lapack_funcs(('potrf',), (self.hess,))
+
+        # Get info about Hessian
+        self.dimension = len(self.hess)
+        self.hess_gershgorin_lb,\
+            self.hess_gershgorin_ub = gershgorin_bounds(self.hess)
+        self.hess_inf = norm(self.hess, np.Inf)
+        self.hess_fro = norm(self.hess, 'fro')
+
+        # A constant such that for vectors smaler than that
+        # backward substituition is not reliable. It was stabilished
+        # based on Golub, G. H., Van Loan, C. F. (2013).
+        # "Matrix computations". Forth Edition. JHU press., p.165.
+        self.CLOSE_TO_ZERO = self.dimension*self.EPS*self.hess_inf
+
+    def _initial_values(self, trust_radius):
```

Comment:
As `trust_radius` is not really a term, I suggest to consider alternative `tr_radius` (at least I decided at some point that it is a better name). But really up to you here.